```
modules/gdscript/language_server/gdscript_language_protocol.cpp: In member function 'void GDScriptLanguageProtocol::stop()':

modules/gdscript/language_server/gdscript_language_protocol.cpp:157:13: error: suggest parentheses around assignment used as truth value [-Werror=parentheses]

  157 |  while (ptr = clients.next(ptr)) {

      |         ~~~~^~~~~~~~~~~~~~~~~~~
```

--- a/modules/gdscript/language_server/gdscript_extend_parser.cpp
+++ b/modules/gdscript/language_server/gdscript_extend_parser.cpp
@@ -157,7 +157,7 @@ void ExtendGDScriptParser::parse_class_symbol(const GDScriptParser::ClassNode *p
 	r_symbol.selectionRange.start.line = r_symbol.range.start.line;
 	r_symbol.detail = "class " + r_symbol.name;
 	bool is_root_class = &r_symbol == &class_symbol;
-	r_symbol.documentation = parse_documentation_as_markdown(is_root_class ? 0 : LINE_NUMBER_TO_INDEX(p_class->line), is_root_class);
+	r_symbol.documentation = parse_documentation(is_root_class ? 0 : LINE_NUMBER_TO_INDEX(p_class->line), is_root_class);
 
 	for (int i = 0; i < p_class->variables.size(); ++i) {
 
@@ -184,7 +184,7 @@ void ExtendGDScriptParser::parse_class_symbol(const GDScriptParser::ClassNode *p
 			symbol.detail += " = " + JSON::print(m.default_value);
 		}
 
-		symbol.documentation = parse_documentation_as_markdown(line);
+		symbol.documentation = parse_documentation(line);
 		symbol.uri = uri;
 		symbol.script_path = path;
 
@@ -204,7 +204,7 @@ void ExtendGDScriptParser::parse_class_symbol(const GDScriptParser::ClassNode *p
 		symbol.range.end.line = symbol.range.start.line;
 		symbol.range.end.character = lines[line].length();
 		symbol.selectionRange.start.line = symbol.range.start.line;
-		symbol.documentation = parse_documentation_as_markdown(line);
+		symbol.documentation = parse_documentation(line);
 		symbol.uri = uri;
 		symbol.script_path = path;
 		symbol.detail = "signal " + signal.name + "(";
@@ -233,7 +233,7 @@ void ExtendGDScriptParser::parse_class_symbol(const GDScriptParser::ClassNode *p
 		symbol.range.end.line = symbol.range.start.line;
 		symbol.range.end.character = lines[line].length();
 		symbol.selectionRange.start.line = symbol.range.start.line;
-		symbol.documentation = parse_documentation_as_markdown(line);
+		symbol.documentation = parse_documentation(line);
 		symbol.uri = uri;
 		symbol.script_path = path;
 
@@ -301,7 +301,7 @@ void ExtendGDScriptParser::parse_function_symbol(const GDScriptParser::FunctionN
 	r_symbol.range.end.line = MAX(p_func->body->end_line - 2, p_func->body->line);
 	r_symbol.range.end.character = lines[r_symbol.range.end.line].length();
 	r_symbol.selectionRange.start.line = r_symbol.range.start.line;
-	r_symbol.documentation = parse_documentation_as_markdown(line);
+	r_symbol.documentation = parse_documentation(line);
 	r_symbol.uri = uri;
 	r_symbol.script_path = path;
 
@@ -359,63 +359,9 @@ void ExtendGDScriptParser::parse_function_symbol(const GDScriptParser::FunctionN
 		if (var->datatype.kind != GDScriptParser::DataType::UNRESOLVED) {
 			symbol.detail += ": " + var->datatype.to_string();
 		}
-		symbol.documentation = parse_documentation_as_markdown(line);
+		symbol.documentation = parse_documentation(line);
 		r_symbol.children.push_back(symbol);
 	}
-}
-
-String ExtendGDScriptParser::marked_documentation(const String &p_bbcode) {
-
-	String markdown = p_bbcode.strip_edges();
-
-	Vector<String> lines = markdown.split("\n");
-	bool in_code_block = false;
-	int code_block_indent = -1;
-
-	markdown = "";
-	for (int i = 0; i < lines.size(); i++) {
-		String line = lines[i];
-		int block_start = line.find("[codeblock]");
-		if (block_start != -1) {
-			code_block_indent = block_start;
-			in_code_block = true;
-			line = "\n";
-		} else if (in_code_block) {
-			line = "\t" + line.substr(code_block_indent, line.length());
-		}
-
-		if (in_code_block && line.find("[/codeblock]") != -1) {
-			line = "\n";
-			in_code_block = false;
-		}
-
-		if (!in_code_block) {
-			line = line.strip_edges();
-			line = line.replace("[code]", "`");
-			line = line.replace("[/code]", "`");
-			line = line.replace("[i]", "*");
-			line = line.replace("[/i]", "*");
-			line = line.replace("[b]", "**");
-			line = line.replace("[/b]", "**");
-			line = line.replace("[u]", "__");
-			line = line.replace("[/u]", "__");
-			line = line.replace("[method ", "`");
-			line = line.replace("[member ", "`");
-			line = line.replace("[signal ", "`");
-			line = line.replace("[enum ", "`");
-			line = line.replace("[constant ", "`");
-			line = line.replace("[", "`");
-			line = line.replace("]", "`");
-		}
-
-		if (!in_code_block && i < lines.size() - 1) {
-			line += "\n\n";
-		} else if (i < lines.size() - 1) {
-			line += "\n";
-		}
-		markdown += line;
-	}
-	return markdown;
 }
 
 String ExtendGDScriptParser::parse_documentation(int p_line, bool p_docs_down) {
@@ -574,10 +520,6 @@ const lsp::DocumentSymbol *ExtendGDScriptParser::search_symbol_defined_at_line(i
 		}
 	}
 	return ret;
-}
-
-String ExtendGDScriptParser::parse_documentation_as_markdown(int p_line, bool p_docs_down) {
-	return marked_documentation(parse_documentation(p_line, p_docs_down));
 }
 
 const lsp::DocumentSymbol *ExtendGDScriptParser::get_symbol_defined_at_line(int p_line) const {

--- a/modules/gdscript/language_server/gdscript_extend_parser.h
+++ b/modules/gdscript/language_server/gdscript_extend_parser.h
@@ -75,11 +75,6 @@ class ExtendGDScriptParser : public GDScriptParser {
 
 	Array member_completions;
 
-	String parse_documentation_as_markdown(int p_line, bool p_docs_down = false);
-
-public:
-	static String marked_documentation(const String &p_bbcode);
-
 public:
 	_FORCE_INLINE_ const String &get_path() const { return path; }
 	_FORCE_INLINE_ const Vector<String> &get_lines() const { return lines; }

--- a/modules/gdscript/language_server/gdscript_language_protocol.cpp
+++ b/modules/gdscript/language_server/gdscript_language_protocol.cpp
@@ -153,9 +153,10 @@ Error GDScriptLanguageProtocol::start(int p_port) {
 }
 
 void GDScriptLanguageProtocol::stop() {
-	const int *ptr = NULL;
-	while (ptr = clients.next(ptr)) {
+	const int *ptr = clients.next(NULL);
+	while (ptr) {
 		clients.get(*ptr)->close();
+		ptr = clients.next(ptr);
 	}
 	server->stop();
 	clients.clear();

--- a/modules/gdscript/language_server/gdscript_text_document.cpp
+++ b/modules/gdscript/language_server/gdscript_text_document.cpp
@@ -39,6 +39,7 @@
 void GDScriptTextDocument::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("didOpen"), &GDScriptTextDocument::didOpen);
 	ClassDB::bind_method(D_METHOD("didChange"), &GDScriptTextDocument::didChange);
+	ClassDB::bind_method(D_METHOD("nativeSymbol"), &GDScriptTextDocument::nativeSymbol);
 	ClassDB::bind_method(D_METHOD("documentSymbol"), &GDScriptTextDocument::documentSymbol);
 	ClassDB::bind_method(D_METHOD("completion"), &GDScriptTextDocument::completion);
 	ClassDB::bind_method(D_METHOD("resolve"), &GDScriptTextDocument::resolve);
@@ -75,6 +76,11 @@ lsp::TextDocumentItem GDScriptTextDocument::load_document_item(const Variant &p_
 	return doc;
 }
 
+void GDScriptTextDocument::notify_client_show_symbol(const lsp::DocumentSymbol *symbol) {
+	ERR_FAIL_NULL(symbol);
+	GDScriptLanguageProtocol::get_singleton()->notify_client("gdscript/show_native_symbol", symbol->to_json(true));
+}
+
 void GDScriptTextDocument::initialize() {
 
 	if (GDScriptLanguageProtocol::get_singleton()->is_smart_resolve_enabled()) {
@@ -100,6 +106,21 @@ void GDScriptTextDocument::initialize() {
 			class_ptr = native_members.next(class_ptr);
 		}
 	}
+}
+
+Variant GDScriptTextDocument::nativeSymbol(const Dictionary &p_params) {
+
+	Variant ret;
+
+	lsp::NativeSymbolInspectParams params;
+	params.load(p_params);
+
+	if (const lsp::DocumentSymbol *symbol = GDScriptLanguageProtocol::get_singleton()->get_workspace()->resolve_native_symbol(params)) {
+		ret = symbol->to_json(true);
+		notify_client_show_symbol(symbol);
+	}
+
+	return ret;
 }
 
 Array GDScriptTextDocument::documentSymbol(const Dictionary &p_params) {
@@ -362,7 +383,7 @@ Array GDScriptTextDocument::definition(const Dictionary &p_params) {
 				}
 				call_deferred("show_native_symbol_in_editor", id);
 			} else {
-				GDScriptLanguageProtocol::get_singleton()->notify_client("gdscript/show_native_symbol", symbol->to_json(true));
+				notify_client_show_symbol(symbol);
 			}
 		}
 	} else if (GDScriptLanguageProtocol::get_singleton()->is_smart_resolve_enabled()) {

--- a/modules/gdscript/language_server/gdscript_text_document.h
+++ b/modules/gdscript/language_server/gdscript_text_document.h
@@ -52,8 +52,10 @@ protected:
 
 private:
 	lsp::TextDocumentItem load_document_item(const Variant &p_param);
+	void notify_client_show_symbol(const lsp::DocumentSymbol *symbol);
 
 public:
+	Variant nativeSymbol(const Dictionary &p_params);
 	Array documentSymbol(const Dictionary &p_params);
 	Array completion(const Dictionary &p_params);
 	Dictionary resolve(const Dictionary &p_params);

--- a/modules/gdscript/language_server/gdscript_workspace.cpp
+++ b/modules/gdscript/language_server/gdscript_workspace.cpp
@@ -475,6 +475,24 @@ void GDScriptWorkspace::resolve_related_symbols(const lsp::TextDocumentPositionP
 	}
 }
 
+const lsp::DocumentSymbol *GDScriptWorkspace::resolve_native_symbol(const lsp::NativeSymbolInspectParams &p_params) {
+
+	if (Map<StringName, lsp::DocumentSymbol>::Element *E = native_symbols.find(p_params.native_class)) {
+		const lsp::DocumentSymbol &symbol = E->get();
+		if (p_params.symbol_name.empty() || p_params.symbol_name == symbol.name) {
+			return &symbol;
+		}
+
+		for (int i = 0; i < symbol.children.size(); ++i) {
+			if (symbol.children[i].name == p_params.symbol_name) {
+				return &(symbol.children[i]);
+			}
+		}
+	}
+
+	return NULL;
+}
+
 void GDScriptWorkspace::resolve_document_links(const String &p_uri, List<lsp::DocumentLink> &r_list) {
 	if (const ExtendGDScriptParser *parser = get_parse_successed_script(get_file_path(p_uri))) {
 		const List<lsp::DocumentLink> &links = parser->get_document_links();

--- a/modules/gdscript/language_server/gdscript_workspace.cpp
+++ b/modules/gdscript/language_server/gdscript_workspace.cpp
@@ -198,7 +198,7 @@ Error GDScriptWorkspace::initialize() {
 		if (!class_data.inherits.empty()) {
 			class_symbol.detail += " extends " + class_data.inherits;
 		}
-		class_symbol.documentation = ExtendGDScriptParser::marked_documentation(class_data.brief_description) + "\n" + ExtendGDScriptParser::marked_documentation(class_data.description);
+		class_symbol.documentation = class_data.brief_description + "\n" + class_data.description;
 
 		for (int i = 0; i < class_data.constants.size(); i++) {
 			const DocData::ConstantDoc &const_data = class_data.constants[i];
@@ -211,7 +211,7 @@ Error GDScriptWorkspace::initialize() {
 				symbol.detail += ": " + const_data.enumeration;
 			}
 			symbol.detail += " = " + const_data.value;
-			symbol.documentation = ExtendGDScriptParser::marked_documentation(const_data.description);
+			symbol.documentation = const_data.description;
 			class_symbol.children.push_back(symbol);
 		}
 
@@ -232,7 +232,7 @@ Error GDScriptWorkspace::initialize() {
 			} else {
 				symbol.detail += ": " + data.type;
 			}
-			symbol.documentation = ExtendGDScriptParser::marked_documentation(data.description);
+			symbol.documentation = data.description;
 			class_symbol.children.push_back(symbol);
 		}
 
@@ -270,7 +270,7 @@ Error GDScriptWorkspace::initialize() {
 			}
 
 			symbol.detail = "func " + class_name + "." + data.name + "(" + params + ") -> " + data.return_type;
-			symbol.documentation = ExtendGDScriptParser::marked_documentation(data.description);
+			symbol.documentation = data.description;
 			class_symbol.children.push_back(symbol);
 		}
 

--- a/modules/gdscript/language_server/gdscript_workspace.h
+++ b/modules/gdscript/language_server/gdscript_workspace.h
@@ -80,7 +80,7 @@ public:
 
 	const lsp::DocumentSymbol *resolve_symbol(const lsp::TextDocumentPositionParams &p_doc_pos, const String &p_symbol_name = "", bool p_func_requred = false);
 	void resolve_related_symbols(const lsp::TextDocumentPositionParams &p_doc_pos, List<const lsp::DocumentSymbol *> &r_list);
-
+	const lsp::DocumentSymbol *resolve_native_symbol(const lsp::NativeSymbolInspectParams &p_params);
 	void resolve_document_links(const String &p_uri, List<lsp::DocumentLink> &r_list);
 	Dictionary generate_script_api(const String &p_path);
 

--- a/modules/gdscript/language_server/lsp.hpp
+++ b/modules/gdscript/language_server/lsp.hpp
@@ -1134,7 +1134,7 @@ struct DocumentSymbol {
 	 */
 	Vector<DocumentSymbol> children;
 
-	_FORCE_INLINE_ Dictionary to_json(bool with_doc = false) const {
+	Dictionary to_json(bool with_doc = false) const {
 		Dictionary dict;
 		dict["name"] = name;
 		dict["detail"] = detail;
@@ -1230,6 +1230,17 @@ struct DocumentSymbol {
 		}
 
 		return item;
+	}
+};
+
+struct NativeSymbolInspectParams {
+
+	String native_class;
+	String symbol_name;
+
+	void load(const Dictionary &p_params) {
+		native_class = p_params["native_class"];
+		symbol_name = p_params["symbol_name"];
 	}
 };
 

--- a/modules/gdscript/language_server/lsp.hpp
+++ b/modules/gdscript/language_server/lsp.hpp
@@ -37,6 +37,9 @@ namespace lsp {
 
 typedef String DocumentUri;
 
+/** Format BBCode documentation from DocData to markdown */
+static String marked_documentation(const String &p_bbcode);
+
 /**
  * Text documents are identified using a URI. On the protocol level, URIs are passed as strings.
  */
@@ -281,6 +284,9 @@ struct Command {
 		return dict;
 	}
 };
+
+// Use namespace instead of enumeration to follow the LSP specifications
+// lsp::EnumName::EnumValue is OK but lsp::EnumValue is not
 
 namespace TextDocumentSyncKind {
 /**
@@ -592,6 +598,7 @@ struct TextDocumentContentChangeEvent {
 	}
 };
 
+// Use namespace instead of enumeration to follow the LSP specifications
 namespace DiagnosticSeverity {
 /**
 	 * Reports an error.
@@ -692,6 +699,7 @@ struct Diagnostic {
 	}
 };
 
+// Use namespace instead of enumeration to follow the LSP specifications
 /**
  * Describes the content type that a client supports in various
  * result literals like `Hover`, `ParameterInfo` or `CompletionItem`.
@@ -756,6 +764,9 @@ struct MarkupContent {
 	}
 };
 
+// Use namespace instead of enumeration to follow the LSP specifications
+// lsp::EnumName::EnumValue is OK but lsp::EnumValue is not
+// And here C++ compilers are unhappy with our enumeration name like Color, File, Reference etc.
 /**
  * The kind of a completion entry.
  */
@@ -787,6 +798,7 @@ static const int Operator = 24;
 static const int TypeParameter = 25;
 }; // namespace CompletionItemKind
 
+// Use namespace instead of enumeration to follow the LSP specifications
 /**
  * Defines whether the insert text in a completion item should be interpreted as
  * plain text or a snippet.
@@ -979,6 +991,9 @@ struct CompletionList {
 	Vector<CompletionItem> items;
 };
 
+// Use namespace instead of enumeration to follow the LSP specifications
+// lsp::EnumName::EnumValue is OK but lsp::EnumValue is not
+// And here C++ compilers are unhappy with our enumeration name like String, Array, Object etc
 /**
  * A symbol kind.
  */
@@ -1181,7 +1196,7 @@ struct DocumentSymbol {
 			markdown.value = "\t" + detail + "\n\n";
 		}
 		if (documentation.length()) {
-			markdown.value += documentation + "\n\n";
+			markdown.value += marked_documentation(documentation) + "\n\n";
 		}
 		if (script_path.length()) {
 			markdown.value += "Defined in [" + script_path + "](" + uri + ")";
@@ -1304,6 +1319,7 @@ struct FoldingRange {
 	}
 };
 
+// Use namespace instead of enumeration to follow the LSP specifications
 /**
  * How a completion was triggered
  */
@@ -1550,6 +1566,61 @@ struct InitializeResult {
 		return dict;
 	}
 };
+
+/** Format BBCode documentation from DocData to markdown */
+static String marked_documentation(const String &p_bbcode) {
+
+	String markdown = p_bbcode.strip_edges();
+
+	Vector<String> lines = markdown.split("\n");
+	bool in_code_block = false;
+	int code_block_indent = -1;
+
+	markdown = "";
+	for (int i = 0; i < lines.size(); i++) {
+		String line = lines[i];
+		int block_start = line.find("[codeblock]");
+		if (block_start != -1) {
+			code_block_indent = block_start;
+			in_code_block = true;
+			line = "\n";
+		} else if (in_code_block) {
+			line = "\t" + line.substr(code_block_indent, line.length());
+		}
+
+		if (in_code_block && line.find("[/codeblock]") != -1) {
+			line = "\n";
+			in_code_block = false;
+		}
+
+		if (!in_code_block) {
+			line = line.strip_edges();
+			line = line.replace("[code]", "`");
+			line = line.replace("[/code]", "`");
+			line = line.replace("[i]", "*");
+			line = line.replace("[/i]", "*");
+			line = line.replace("[b]", "**");
+			line = line.replace("[/b]", "**");
+			line = line.replace("[u]", "__");
+			line = line.replace("[/u]", "__");
+			line = line.replace("[method ", "`");
+			line = line.replace("[member ", "`");
+			line = line.replace("[signal ", "`");
+			line = line.replace("[enum ", "`");
+			line = line.replace("[constant ", "`");
+			line = line.replace("[", "`");
+			line = line.replace("]", "`");
+		}
+
+		if (!in_code_block && i < lines.size() - 1) {
+			line += "\n\n";
+		} else if (i < lines.size() - 1) {
+			line += "\n";
+		}
+		markdown += line;
+	}
+	return markdown;
+}
 
 } // namespace lsp
 

